File: src/diagnostics/output.zig
const std = @import("std");
const Counter = @import("counter.zig").Counter;
const Event = @import("event.zig").Event;

/// prints the summary for the diagnostics
/// TODO:
/// 1. make it read from the file instead
/// 2. make it verbose and easy to understand by the developer
pub fn print_summary(counter: Counter) void {
    inline for (@typeInfo(Event).Enum.fields) |field| {
        const event: Event = @enumFromInt(field.value);
        const count = counter.get(event);
        if (count > 0) {
            std.debug.print("{s}: {d}\n", .{ field.name, count });
        }
    }
}

/// writes all the diagnostics to the file : `cache/diagnostics.txt`
pub fn write_to_file(counter: Counter, path: []const u8) !void {
    const file = try std.fs.cwd().createFile(path, .{});
    defer file.close();

    var buf = std.io.bufferedWriter(file.writer());
    const writer = buf.writer();

    inline for (@typeInfo(Event).Enum.fields) |field| {
        const event: Event = @enumFromInt(field.value);
        const count = counter.get(event);
        try writer.print("{s}: {d}\n", .{ field.name, count });
    }
    try buf.flush();
}

File: src/diagnostics/output.cpp
#include "output.hpp"

#include <fstream>
#include <iostream>

namespace Pico {

void print_summary(const Counter &counter) {
  for (size_t i = 0; i < EVENT_COUNT; ++i) {
    auto event = static_cast<Event>(i);
    uint32_t count = counter.get(event);
    if (count > 0) {
      std::cout << event_name(event) << ": " << count << "\n";
    }
  }
}

int write_to_file(const Counter &counter, const std::string &path) {
  std::ofstream file(path);
  if (!file.is_open()) {
    return -1;
  }

  for (size_t i = 0; i < EVENT_COUNT; ++i) {
    auto event = static_cast<Event>(i);
    uint32_t count = counter.get(event);
    file << event_name(event) << ": " << count << "\n";
  }

  file.flush();
  return 0;
}

} // namespace Pico
